pkg/sql/schemachanger/sctest/cumulative.go

```diff
@@ -52,34 +481,66 @@ import (
 func cumulativeTest(
 	t *testing.T,
 	relPath string,
-	tf func(t *testing.T, path string, rewrite bool, setup, stmts []parser.Statement),
+	tf func(t *testing.T, path string, rewrite bool, setup, stmts []parser.Statement, stageExecMap *stageExecStmtMap),
 ) {
 	skip.UnderStress(t)
 	skip.UnderRace(t)
 	path := testutils.RewritableDataPath(t, relPath)
 	var setup []parser.Statement
-	datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
-		stmts, err := parser.Parse(d.Input)
-		require.NoError(t, err)
-		require.NotEmpty(t, stmts)
+	stageExecMap := makeStageExecStmtMap()
+	rewrite := false
+	var testStmts parser.Statements
+	var lines []string
+	numTestStmts := 0
 
+	// First pass collect information.
+	datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
+		require.Zero(t, numTestStmts, "only one test command per-test, "+
+			"and it must be the last one.")
 		switch d.Cmd {
 		case "setup":
-			// no-op
+			stmts, err := parser.Parse(d.Input)
+			setup = append(setup, stmts...)
+			require.NoError(t, err)
+			require.NotEmpty(t, stmts)
+		// no-op
+		case "stage-exec":
+			// DML injected statements will only be executed on cumalative tests,
+			// for end-to-end tests these are fully ignored.
+			stageExecMap.ParseStageExec(t, d)
+		case "stage-query":
+			stageExecMap.ParseStageQuery(t, d)
 		case "test":
-			var lines []string
+			stmts, err := parser.Parse(d.Input)
+			require.NoError(t, err)
+			require.NotEmpty(t, stmts)
+			testStmts = stmts
 			for _, stmt := range stmts {
 				lines = append(lines, stmt.SQL)
 			}
-			t.Run(strings.Join(lines, "; "), func(t *testing.T) {
-				tf(t, path, d.Rewrite, setup, stmts)
-			})
-		default:
-			return fmt.Sprintf("unknown command: %s", d.Cmd)
+			rewrite = d.Rewrite
+			numTestStmts += 1
+			tf(t, path, rewrite, setup, testStmts, stageExecMap)
 		}
-		setup = append(setup, stmts...)
 		return d.Expected
 	})
+	// Run through and recover the observed output for statements.
+	if rewrite {
+		datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
+			switch d.Cmd {
+			// Retrieve the generated output from the previous execution in the
+			// rewrite mode of DML injection. We are going to the store the
+			// observed output based on the line number and file names.
+			case "stage-exec":
+				fallthrough
+			case "stage-query":
+				return stageExecMap.GetExpectedOutputForPos(d.Pos)
+			}
+			// For all other commmands the output will be generated in the end-to-end
+			// test.
+			return d.Expected
```

Comment:
okay, so we have essentially two layers here -- end-to-end and cumulative. Cumulative is the foundational layer that all the "high-level" testing will invoke: end-to-end, rollback, pause-resume, backup-restore. You put this `stage-exec`/`stage-query` logic here in the cumulative layer, and that means we are going to inject stmts in all the "high-level" testings, is this right? (if so, that'd be great! It's so cool to say that "we are injecting DML stms while running a DDL statement while doing backup/restore)

By putting the rewrite logic here, you also say "the rewrite for `stage-query` is responsible in the cumulative layer, while the rewrite for the `test` stmt in the testdata file is responsible in the end-to-end layer". Is this right? If so, I think this makes sense since the expected output for the `test` command is only used/concerned in the end-to-end test.

Reply:
So, for backup and restore we don't have an injection because the callback and map are not used. Think we could potentially add support in a later PR, but coverage wise it probably won't buy us much.

I added comments in both places for cumulative:
```
			// cumlativeTest will rewrite only the stage-exec and stage-query commands,
			// all others are rewritten by the end-to-end tests.
```

and end-to-end
```
			// Both of these commands are DML injections, which is not relevant
			// for end-to-end testing. We don't actually execute statements here.
			// So return the original output, these will get rewritten in
			// cumulativeTest.
```